### PR TITLE
K8SPXC-485 add ability to run logcollector as a sidecar for pxc pod

### PR DESCRIPTION
**Pull request by hors**
[K8SPXC-485](https://jira.percona.com/browse/K8SPXC-485)

### CODE REVIEW

File: deploy/cr.yaml [hunk 1]
```diff
@@ -309,6 +309,18 @@ spec:
 #     - 10.0.0.0/8
 #   serviceAnnotations:
 #     service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
+  logcollector:
+    enabled: false
+    image: percona/percona-xtradb-cluster-operator:1.6.0-fluentbit
```
Comment (delgod) [inline deploy/cr.yaml:314]:
```suggestion
    image: percona/percona-xtradb-cluster-operator:1.6.0-logcollector
```

File: pkg/controller/pxc/controller.go [hunk 2]
```diff
@@ -868,6 +868,23 @@ func (r *ReconcilePerconaXtraDBCluster) reconcileConfigMap(cr *api.PerconaXtraDB
 		}
 	}
 
+	if cr.Spec.LogCollector != nil && cr.Spec.LogCollector.Configuration != "" && cr.CompareVersionWith("1.7.0") >= 0 {
+		configMap := config.NewConfigMap(cr, ls["app.kubernetes.io/instance"]+"-logcollactor", "fluentbit_custom.conf", cr.Spec.LogCollector.Configuration)
```
Comment (delgod) [inline pkg/controller/pxc/controller.go:872]:
```suggestion
		configMap := config.NewConfigMap(cr, ls["app.kubernetes.io/instance"]+"-logcollector", "fluentbit_custom.conf", cr.Spec.LogCollector.Configuration)
```

File: pkg/pxc/app/statefulset/node.go [hunk 3]
```diff
@@ -289,6 +359,9 @@ func (c *Node) Volumes(podSpec *api.PodSpec, cr *api.PerconaXtraDBCluster) (*api
 			vol.Volumes,
 			app.GetSecretVolumes(VaultSecretVolumeName, podSpec.VaultSecretName, true))
 	}
+	if cr.Spec.LogCollector != nil && cr.Spec.LogCollector.Configuration != "" && cr.CompareVersionWith("1.7.0") >= 0 {
+		vol.Volumes = append(vol.Volumes, app.GetConfigVolumes("custom-config", ls["app.kubernetes.io/instance"]+"-logcollactor"))
```
Comment (delgod) [inline pkg/pxc/app/statefulset/node.go:363]:
```suggestion
		vol.Volumes = append(vol.Volumes, app.GetConfigVolumes("custom-config", ls["app.kubernetes.io/instance"]+"-logcollector"))
```

File: pkg/pxc/app/statefulset/node.go [hunk 4]
```diff
@@ -211,6 +225,62 @@ func (c *Node) SidecarContainers(spec *api.PodSpec, secrets string, cr *api.Perc
...
+	container.Resources = res
+
+	container.VolumeMounts = []corev1.VolumeMount{
+		{
+			Name:      DataVolumeName,
+			MountPath: "/var/lib/mysql",
+		},
+	}
+
+	if cr.Spec.LogCollector != nil && cr.Spec.LogCollector.Configuration != "" {
```
Comment (delgod) [inline pkg/pxc/app/statefulset/node.go:274]: can it be a secret?

File: deploy/cr.yaml [hunk 5]
```diff
@@ -309,6 +309,18 @@ spec:
...
+#    configuration: |
+#      [INPUT]
+#          Name  tail
+#          Path   ${LOG_DATA_DIR}/innobackup.backup.log
+#          Tag   ${POD_NAMESPASE}.${POD_NAME}.innobackup.backup.log
```
Comment (delgod) [inline deploy/cr.yaml:319]: it is better to add an example for elastic search output

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| haproxy | passed |
| storage | failed |
| big-data | passed |
| recreate | passed |
| smart-update | failed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring | failed |
| demand-backup | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| affinity | failed |
| self-healing-advanced | passed |
| one-pod | passed |
| auto-tuning | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| operator-self-healing | passed |
| demand-backup-encrypted-with-tls | passed |

File: deploy/cr.yaml (hunk 1 above)
Comment (hors) [reply deploy/cr.yaml:315]: done

File: pkg/controller/pxc/controller.go (hunk 2 above)
Comment (hors) [reply pkg/controller/pxc/controller.go:872]: done

File: pkg/pxc/app/statefulset/node.go (hunk 3 above)
Comment (hors) [reply pkg/pxc/app/statefulset/node.go:374]: done

File: pkg/pxc/app/statefulset/node.go (hunk 4 above)
Comment (hors) [reply pkg/pxc/app/statefulset/node.go:285]: we will use secret for env variables which can be used as a secrets

File: deploy/cr.yaml (hunk 5 above)
Comment (hors) [reply deploy/cr.yaml:320]: done

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | failed |
| scaling-proxysql | passed |
| self-healing | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| self-healing-advanced | failed |
| security-context | passed |
| operator-self-healing | passed |

File: deploy/cr.yaml [hunk 6]
```diff
@@ -309,6 +310,21 @@ spec:
 #     - 10.0.0.0/8
 #   serviceAnnotations:
 #     service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
+  logcollector:
+    enabled: false
```
Comment (delgod) [inline deploy/cr.yaml:314]: should be enabled by default according to spec

**Review by delgod — CHANGES_REQUESTED** [review]

File: deploy/cr.yaml (hunk 6 above)
Comment (hors) [reply deploy/cr.yaml:314]: I suggest to enable it as soon as we have log rotation

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | failed |
| scaling | passed |
| smart-update | failed |
| haproxy | passed |
| recreate | passed |
| big-data | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| upgrade | passed |
| monitoring | failed |
| demand-backup | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| affinity | failed |
| security-context | passed |
| self-healing-advanced | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| demand-backup-encrypted-with-tls | passed |

File: deploy/cr.yaml (hunk 6 above)
Comment (hors) [reply deploy/cr.yaml]: it was done

File: pkg/pxc/app/statefulset/node.go [hunk 7]
```diff
@@ -211,6 +225,103 @@ func (c *Node) SidecarContainers(spec *api.PodSpec, secrets string, cr *api.Perc
...
+				},
+			},
+		},
+	}
+
+	logRotEnvs := []corev1.EnvVar{
+		{
+			Name:  "PROCESS_NAME",
+			Value: "logrotate",
+		},
```
Comment (delgod) [inline pkg/pxc/app/statefulset/node.go:261]: why it is needed? please use `corev1.Container.Args` option instead

File: pkg/pxc/app/statefulset/node.go [hunk 8]
```diff
@@ -211,6 +225,103 @@ func (c *Node) SidecarContainers(spec *api.PodSpec, secrets string, cr *api.Perc
...
+		Resources:       res,
+		VolumeMounts: []corev1.VolumeMount{
+			{
+				Name:      DataVolumeName,
+				MountPath: "/var/lib/mysql",
+			},
+		},
+	}
+
+	if cr.Spec.LogCollector != nil && cr.Spec.LogCollector.Configuration != "" {
```
Comment (delgod) [inline pkg/pxc/app/statefulset/node.go:315]: how to pass custom logrotate config?

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| big-data | passed |
| smart-update | failed |
| haproxy | passed |
| recreate | passed |
| storage | failed |
| scaling-proxysql | passed |
| init-deploy | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing | passed |
| upgrade-consistency | passed |
| monitoring | failed |
| demand-backup | passed |
| self-healing-advanced | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| affinity | failed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| scheduled-backup | passed |
| users | failed |
| demand-backup-encrypted-with-tls | passed |

File: pkg/pxc/app/statefulset/node.go [hunk 9]
```diff
@@ -211,6 +225,103 @@ func (c *Node) SidecarContainers(spec *api.PodSpec, secrets string, cr *api.Perc
...
+			Name: "MONITOR_PASSWORD",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: app.SecretKeySelector(logRsecrets, "monitor"),
+			},
+		},
+	}
+
+	fvar := true
+	logProcContainer := corev1.Container{
+		Name:            "pxc-logcollector",
```
[the remark this answers is not in the document]
Comment (hors) [reply pkg/pxc/app/statefulset/node.go]: fixed

File: pkg/pxc/app/statefulset/node.go [hunk 10]
```diff
@@ -211,6 +225,103 @@ func (c *Node) SidecarContainers(spec *api.PodSpec, secrets string, cr *api.Perc
...
+		VolumeMounts: []corev1.VolumeMount{
+			{
+				Name:      DataVolumeName,
+				MountPath: "/var/lib/mysql",
+			},
+		},
+	}
+
+	logRotContainer := corev1.Container{
+		Name:            "pxc-logrotate",
```
[the remark this answers is not in the document]
Comment (hors) [reply pkg/pxc/app/statefulset/node.go]: fixed

**Review by delgod — DISMISSED** [review]

**Review by qJkee — DISMISSED** [review]

File: pkg/pxc/app/statefulset/node.go [hunk 11]
```diff
@@ -211,6 +225,102 @@ func (c *Node) SidecarContainers(spec *api.PodSpec, secrets string, cr *api.Perc
...
+			{
+				Name:      DataVolumeName,
+				MountPath: "/var/lib/mysql",
+			},
+		},
+	}
+
+	if cr.Spec.LogCollector != nil && cr.Spec.LogCollector.Configuration != "" {
+		logProcContainer.VolumeMounts = append(logProcContainer.VolumeMounts, corev1.VolumeMount{
+			Name:      "custom-config",
```
Comment (delgod) [inline pkg/pxc/app/statefulset/node.go:316]: can we please rename?
```suggestion
			Name:      "logcollector-config",
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/pxc/app/statefulset/node.go (hunk 11 above)
Comment (hors) [reply pkg/pxc/app/statefulset/node.go]: done

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| recreate | passed |
| upgrade-haproxy | failed |
| big-data | passed |
| scaling | passed |
| storage | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| upgrade-proxysql | failed |
| security-context | passed |
| init-deploy | passed |
| self-healing-advanced | passed |
| limits | failed |
| restore-to-encrypted-cluster | passed |
| smart-update | passed |
| operator-self-healing | failed |
| monitoring | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | failed |
| auto-tuning | passed |
| upgrade-consistency | passed |
| proxysql-sidecar-res-limits | passed |
| scheduled-backup | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| haproxy | passed |
| storage | passed |
| upgrade-haproxy | failed |
| big-data | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| upgrade-proxysql | failed |
| security-context | passed |
| limits | failed |
| smart-update | passed |
| monitoring | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| operator-self-healing | failed |
| affinity | passed |
| one-pod | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| upgrade-consistency | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Review by spron-in — APPROVED** [review]
lgtm

**Review by qJkee — APPROVED** [review]

**Pull request merged**